src/main/java/org/reaktivity/nukleus/amqp/internal/stream/AmqpServerFactory.java

```diff
@@ -326,163 +983,392 @@ private void onNetwork(
+            long traceId,
+            long authorization)
+        {
+            cleanupBudgetCreditorIfNecessary();
+            cleanupEncodeSlotIfNecessary();
+            doEnd(network, routeId, replyId, traceId, authorization, EMPTY_OCTETS);
+        }
+
+        private void doNetworkAbort(
+            long traceId,
+            long authorization)
+        {
+            cleanupBudgetCreditorIfNecessary();
+            cleanupEncodeSlotIfNecessary();
+            doAbort(network, routeId, replyId, traceId, authorization, EMPTY_OCTETS);
+        }
+
+        private void doNetworkReset(
+            long traceId,
+            long authorization)
+        {
+            cleanupDecodeSlotIfNecessary();
+            doReset(network, routeId, initialId, traceId, authorization, EMPTY_OCTETS);
+        }
+
+        private void doNetworkWindow(
+            long traceId,
+            long authorization,
+            int credit,
+            int padding,
+            long budgetId)
+        {
+            assert credit > 0;
+
+            initialBudget += credit;
+            doWindow(network, routeId, initialId, traceId, authorization, budgetId, credit, padding);
+        }
+
+        private void doNetworkSignal(
+            long traceId)
+        {
+            doSignal(network, routeId, initialId, traceId);
+        }
+
+        private void decodeNetwork(
+            long traceId,
+            long authorization,
+            long budgetId,
+            DirectBuffer buffer,
+            int offset,
+            int limit)
+        {
+            AmqpServerDecoder previous = null;
+            int progress = offset;
+            while (progress <= limit && previous != decoder)
+            {
+                previous = decoder;
+                progress = decoder.decode(this, traceId, authorization, budgetId, buffer, progress, limit);
+            }
+
+            if (progress < limit)
+            {
+                if (decodeSlot == NO_SLOT)
                 {
-                    decodeState = this::decodeFrame;
+                    decodeSlot = bufferPool.acquire(initialId);
+                }
+
+                if (decodeSlot == NO_SLOT)
+                {
+                    cleanupNetwork(traceId, authorization);
                 }
                 else
                 {
-                    decodeState = this::decodeInvalid;
-                    doEnd();
+                    final MutableDirectBuffer slotBuffer = bufferPool.buffer(decodeSlot);
+                    decodeSlotLimit = limit - progress;
+                    slotBuffer.putBytes(0, buffer, progress, decodeSlotLimit);
                 }
             }
+            else
+            {
+                cleanupDecodeSlotIfNecessary();
+            }
+        }
+
+        private void onDecodeHeader(
+            long traceId,
+            long authorization,
+            AmqpProtocolHeaderFW header)
+        {
+            if (isAmqpHeaderValid(header))
```

Comment:
What if the header is not valid?

Reply:
Added an `else` statement with `onDecodeError` .